### ISSUE TITLE
Some Zwave devices never receive value updates

### ISSUE THREAD

**Issue by rrozema** [domoticz/domoticz#3727]
This is actually a bug in openzwave, but since the code containing the bug is copied from openzwave into the domoticz codebase I still report it here. I don't have any status in either project and I know some developers here do have access to both project teams (@petergebruers ). Steps that need to be performed to get the fix implemented are:
- openzwave code needs to be fixed (valueID.h)
- the fixed valueID.h needs to be copied over from openzwave into domoticz
- the openzwave library/dll needs to be updated within domoticz
- incorrectly working openzwave nodes need to be re-included in domoticz.

The bug is that there are 2 methods GetValueStoreKey(...) in valueID.h to retrieve a ValueStoreKey. In the current implementation these 2 methods do not return the same value for one and the same device, making that depending on the version of GetValueStoreKey got called, a different value is found (or not found) in a node's ValueStore.

Here is an excerpt from valueID.h:
```
			// Get the key from our own m_id
			uint32 GetValueStoreKey() const
			{
				/* 0xIIIICCii
				 * I = Index
				 * C = CC
				 * i = Instance
				 */
				/*                  CC                             Index                       Instance */
				return (((m_id & 0x003fc000) >> 6) | (m_id1 & 0xffff0000) | ((m_id & 0xFF0) >> 4));
			}

			// Generate a key from its component parts
			static uint32 GetValueStoreKey(uint8 const _commandClassId, uint8 const _instance, uint16 const _valueIndex)
			{

				uint32 key = (((uint32) _instance)) | (((uint32) _commandClassId) << 8) | (((uint32) (_valueIndex & 0xFFFF)) << 16);

				return key;
			}

```

I do not have the experience nor the patience to properly code those bit shifts and masks in the 1st method, so I simply made the change as follows:

```
			// Get the key from our own m_id
			uint32 GetValueStoreKey() const
			{
				return this->GetValueKeyStore( this->GetCommandClass(), this->GetInstance(), this->GetIndex());
			}

```
I had to type this from memory because I'm at work currently, with no access to my development environment at home, so this can contain some typos. But the idea should be clear.

Then I had to rebuild all of openzwave and domoticz and after that my neo coolcam pir sensor suddenly received updates for both their motion device and their lux device, where before this change both got one update upon initial inclusion of the node and after that only the lux device ever received updates (and the other device was shown red in the interface because no updates were received for a long time). The code in openzwave does not always replay events in the exact same order (another bug I'm trying to find), which makes that some devices, even thouigh they are of the same types, behave differently: Depending on which device was added first into the ValueStore, this device does receive updates and updates for the 2nd device get rejected and thus the device never receives any updates any more. The order in which the devices are added is not fixed. For some devices to be re-activated the zwave node needs to be re-included before the proper value gets added into the ValueStore.

**Comment by petergebruers** [comment]
@rrozema you've got my attention.

The bit shuffling makes my head hurt too 😉

I trust you are right about this and I can look into this and possibly make a PR on OZW repo.

By some coincidence, I am adding tests to OZW and the one test suite (the only existing one) I am looking at is about "ValueID". There are no tests on GetValueStoreKey 🙁 

> The order in which the devices are added is not fixed.

Not sure that is a bug. I don't think OZW guarantees order. Haven't looked into that in great detail. But. The node list, read from the from the controller, is fixed as you expect, and OZW also reads your cache file "as is"... but after that OZW also "interviews" devices and depending on the response timing that may change order of events. Does this explanation help?

**Comment by petergebruers** [comment]
Let's find out if the overloads return the same store ID... We'll have to do that bit shuffling.

```cpp
static uint32 GetValueStoreKey(uint8 const _commandClassId, uint8 const _instance, uint16 const _valueIndex)
{

	uint32 key = (((uint32) _instance)) | (((uint32) _commandClassId) << 8) | (((uint32) (_valueIndex & 0xFFFF)) << 16);

	return key;
}
```
So _instance is the easiest... It sits at the bottom 8 bits.
Then Command Class occupies bits 8-15 (0-based count)
And _valueIndex takes 16 to 31

Is this the same as: return (((m_id & 0x003fc000) >> 6) | (m_id1 & 0xffff0000) | ((m_id & 0xFF0) >> 4));

The private data has:

m_id contains (((uint32) _commandClassId) << 14)

So to get the command class from the private data we have to mask with

hex(255<<14)
= 0x3FC000

But then the bits are still in 14-21 -> we need them in 8-15 this means shift them >> 6

m_id also contains (((uint32) (_instance & 0xFF)) << 4)

So to get the instance from private data we first have to mask with

hex(255<<4)
= 0xFF0

But then the bits are still in 4-11 -> we need them in 0-7 this means shift them >> 4

Lastly, the index is in:

m_id1 = (((uint32) _valueIndex) << 16);

The bits are in position 16-31

We have to mask m_id with this to get the index only.

hex(0xFFFF <<16)
= 0xFFFF0000

We want the bits in position 16-31 - no shift here.

Conclusion: both functions return identical results.

I added a test... Not exhaustive but this passes:

```
TEST(ValueID, GetStoreKey)
{
	ValueID *vid1 = new ValueID(0xFFFFu, 0x1, ValueID::ValueGenre_Basic, 0xCC, 0x02, 0x04, ValueID::ValueType_BitSet);
	ValueID *vid2 = new ValueID(0xFFFFu, 0x1, ValueID::ValueGenre_Basic, 0x01, 0x01, 0x01, ValueID::ValueType_BitSet);
	ValueID *vid3 = new ValueID(0xFFFFu, 0x1, ValueID::ValueGenre_Basic, 0x80, 0x80, 0x80, ValueID::ValueType_BitSet);

	EXPECT_EQ(vid1->GetValueStoreKey(), vid1->GetValueStoreKey(0xCC, 0x02, 0x04));
	EXPECT_EQ(vid2->GetValueStoreKey(), vid2->GetValueStoreKey(0x01, 0x01, 0x01));
	EXPECT_EQ(vid3->GetValueStoreKey(), vid3->GetValueStoreKey(0x80, 0x80, 0x80));
	EXPECT_NE(vid1->GetValueStoreKey(), vid2->GetValueStoreKey());
	EXPECT_NE(vid1->GetValueStoreKey(), vid3->GetValueStoreKey());
	EXPECT_NE(vid2->GetValueStoreKey(), vid3->GetValueStoreKey());
	delete vid1;
	delete vid2;
	delete vid3;
}
```

Can you find an example of 3 values that would not behave properly? I can't.

But you clearly saw "something"... not sure what it was.

Please note that GetValueStoreKey is not used by Domoticz - it is only used by OZW internally.

**Comment by gizmocuz** [comment]
@rrozema , could you provide us with the details of the fields for your Motion and Lux fields ?

- commandClass
- instance
- index

Could you take them from your cache file "Before" and "After" you made the changes ?

**Comment by petergebruers** [comment]
I've added the tests I've mentioned and some more to to OpenZWave Version 1.6-954-gaa29b66e - there are no functional changes. If you build from source, "make test" should pass on your system, see cpp/test/ValueID_test.cpp for the actual comparisons

If it does not pass or you've find some combination that would lead both functions to cause a different result please try to add a test, I'll investigate.

There are no functional changes, there are also no code changes to those functions in this commit.

@gizmocuz thanks for chiming in.

@rrozema also please check if a debug version of OZW_Log.txt shows the device sends data to your controller or not. Leave "debug logging" on for a while, we may need info from it anyway. If your device is a Neo NAS-PD02Z I might have spare, but cannot find it at the moment 🙁. It might not be exactly the same though, some version report temp while others don't.

**Comment by rrozema** [comment]
I will switch on debug logging again. My network contains 50+ devices though, so that is going to be a big log file ;-).

I still have 2 sensors that are non-operational (i.e. unmodified), mostly because they're in hard to get to locations like at the top of a stairway. So I've got 2 tries left to see what changes. I will try to collect as much information from one sensor, so that I can try one more time if I mess up the test. I will do the test in 3 phases:

Phase 1: I will switch on openzwave debugging, start the system and wait for all devices to have sent their NIF's (openzwave is initialised). I will trigger the sensor both by waving my hand in front of the sensor and putting it in bright light or shielding it from light. Then I will shutdown and save a copy of the ozw_log.txt. 

Phase 2: Next I will start again, wait for ozw to initialise. Once that is done,  Then re-include the the sensor (without excluding), shutdown and save a copy of ozw_log.txt. 

In phase 3 I will startup again and wait for ozw to initialise. Then I will exclude the sensor, then include it. I will trigger the sensor both by waving my hand in front of the sensor and putting it in bright light or shielding it from light. Then I will shutdown and save the last copy of ozw_log.txt.

What is your opinion, do I -at any point of these steps- need to wake-up the device by pressing the button inside the device?

**Comment by Olifant1990** [comment]
Could this be the same problem as I have with a BeNext P1 dongle?
Only a few "devices" are giving updates. Like I am missing some instances/indexes from a specific command class?

**Comment by rrozema** [comment]
@Olifant1990 maybe. We'll see. 
@petergebruers The sensors I'm testing with is the NAS-PD01Z. This is the one with the holder, not the one with the magnets. The NAS-PD02Z indeed has an additional temperature device, but my PD-01Z's only have 3 devices: "Motion Sensor","Illuminance" and "Home Security". I also have other types nodes that show 'stale' device values (everspring ST814, fibaro fibaro dimmer 2 FGD-212, fibaro double switch 2 FDS-223). But these PD01Z's are most annoying as the motion sensor is the reason these devices are installed, so if this doesn't work, the node is pretty useless. B.t.w. I have 2 systems: a production system (RPi3b with z-way shield), plus a test system (windows 7, visual studio 2019 with debugging capability with aeotec z-stick). Lately both systems are built from latest dev branch domoticz and latest dev branch open-zwave (I delete openzwave header files from the domoticz source tree to make sure I always have the version of these files that matches the open-zwave sources. I normally have no nodes in the test  system. I test specific nodes by 'borrowing' them from the production system. I will today try to get an OZW_log.txt from the test system with a single PD01Z included. But I will also collect ozw_log.txt from the production system before excluding the (malfunctioning) node from there.

**Comment by gizmocuz** [comment]
@rrozema , while your testing , could you also provide the values i asked above ?
You should be fine leaving the headers inside domoticz

**Comment by robgeerts** [comment]
Does this issue have to do with the bug that a device (door sensor in my case) is not being updated in Devices-list (while it is updated in zwave-nodes-list)?

**Comment by Olifant1990** [comment]
And for example, yesterday I rebooted domoticz and my gas meter, which was on ID 0801, was changed to 0804... Is this due to this GetValueStoreKey stuff?
See:
![afbeelding](https://user-images.githubusercontent.com/25199588/69528467-8e832500-0f6e-11ea-9056-01f87d438e04.png)

Maybe also the fact that all the IDs have different notations?
Can we make it uniform?
![afbeelding](https://user-images.githubusercontent.com/25199588/69528556-bf635a00-0f6e-11ea-8085-34028384596e.png)

Or is this something eles then this? Because on the place of C, it is not CC, but my node id here
0xIIIICCii
				 * I = Index
				 * C = CC
				 * i = Instance
				 */

Also, after the reboot, a few elements of the command class which didn't work for the past few weeks, started to update suddenly.
